### ISSUE TITLE
Clipboard API

### ISSUE THREAD

**Issue by TatuLund** [vaadin/flow#17703]
### Describe your motivation

Requests related different clipboard actions are relatively common by our community. Currently Flow lacks easy API for this. It is however possible to implement clipboard actions using Element API and JavaScript using regular Browser API's.

Typical examples are like these

1. Copy text from clipboard -> https://cookbook.vaadin.com/copy-from-clipboard

2. Copy text to clipboard -> https://cookbook.vaadin.com/copy-to-clipboard

3. Paste image from clipboard -> https://gist.github.com/TatuLund/7a5d38f6bb89bb764eaf677e25364524

I assume these are the most common ones as code examples do exists for those due popular demand. Naturally there other analogous cases like copying image to clipboard, or instead text of image handle any common file type. 

### Describe the solution you'd like

Easy to use Clipboard API in Flow that can handle the most common use cases.

### Describe alternatives you've considered

Implementing by Element API and JavaScript

**Comment by tepi** [comment]
First example "Copy text from clipboard" does not work on Firefox or Safari on macOS - seemingly nothing happens, a permission prompt is not produced. On Chrome a permission is asked for and once it's given, copying works.

If this can't be made to function on Safari/Firefox, we might not want to add it to Flow.

**Comment by mshabarov** [comment]
For `readText()` in particular, Safari says it supports it. Firefox doesn't (only in browser's extensions).
See https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API

**Comment by tepi** [comment]
Yes, noticed that as well, but I can't trigger the prompt in Safari (at least on localhost). I get this in the console:

**Comment by mshabarov** [comment]
I got the same, Safari gives me
```
Unhandled Promise Rejection: NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.
```
and FireFox does nothing silently.

Probably that's only for Mac, need to test.

Anyway, I would propose to not implement this feature until `navigator.clipboard` browser API is supported and works in all Vaadin supported browsers without user interaction. Or at least until we realise what a workaround would be (e.g. how to avoid permission rejection).

**Comment by TatuLund** [comment]
![image](https://github.com/user-attachments/assets/92e894bc-a5c0-43c8-9c10-f9950b5483ee)

**Comment by bsutton** [comment]
is there any update on this. I have customers complaining about the problem.

**Comment by mshabarov** [comment]
Makes sense to revisit this feature, just made a quick test with Firefox and the recipes and Clipboard read/write examples seem to work.

We still have to test it with browsers that Vaadin supports and figure out the Java API design. Also what we could include besides basic read/write operations.

**Comment by TatuLund** [comment]
See also: https://vaadin.com/forum/t/clipboard-copy/164697/3

**Comment by mshabarov** [comment]
Firefox 135.0 forces me to click on the "Paste" context menu before I paster from clipboard.

Safari 18.3 shows me "Unhandled Promise Rejection: NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission." console error, apparently because it disables clipboard feature by default, because if I run it manually, I got an error as well:
```javascript
navigator.clipboard.writeText("Test from Safari").then(() => {
    console.log("Clipboard access works!");
}).catch(err => {
    console.error("Clipboard error:", err);
});
```

Vaadin Directory has few add-on for Clipboard that support latest V24 versions, see for example this https://vaadin.com/directory/component/clipboard-for-flow .

I'd de-prioritise this for now until we get enough feature requests or votes to embed it into framework.

**Comment by cstuder** [comment]
Where can I vote or request this as a feature?

**Comment by mshabarov** [comment]
> Where can I vote or request this as a feature?

You can give 👍 to this ticket, that would be enough :)

**Comment by gmerton** [comment]
Strong +1.  It's embarrassing to tell customers that the framework we use can't support copy-to-clipboard functionality.  The next obvious question they ask is: "Why would you use that framework?"

**Comment by mcollovati** [comment]
The Clipboard API proposed in #23615 is well-structured and covers the most common use cases, clearly separating (and documenting limitation) of client-side operations (`copyOnClick`) from server-initiated ones (`writeText`, `readText`).

A few areas could improve the developer experience.

### Error handling

Several operations silently swallow failures. `readText` has no error callback, so permission denials go unnoticed.
The basic `copyOnClick` variant doesn't catch rejected promises, leading to unhandled rejections in the browser console.
The `PastListener` never fires if a file upload fails during a paste event.

### Callback inconsistency

The API mixes different error-handling styles: `copyOnClick` uses `Command` (no error message), `writeText` uses `SerializableConsumer<String>` (with the error message).
The `copyOnClick` variants for source components and images have no callbacks at all. 
A more uniform approach would make the API easier to learn and use.

### Misc

`writeImage` should use `DownloadHandler` instead of the deprecated `StreamResource`.

The `copyOnClick(Component, Component)` hard-codes reading `value` property with fallback to `textContent`. I wonder if there might be situations where a different property should be used and could be configurable through an API parameter.

Paste listener forces preventDefault. This breaks the normal paste-into-field behavior, which could be a surprising default. Developers who just want to observe paste events (e.g., for analytics or validation) shouldn't have to work around the framework eating the event. Making this configurable would be more flexible. 
In addition, `copy` and `cut` event handling should be refactored using `document.getSelection()` on the client side, otherwise data is not forwarded to the server (`event.clipboardData.getData(...)` is for `paste` event only).

**Comment by rucko24** [comment]
Hi, 

This branch I have here is using this new feature of the ClipboardAPI.

```xml
<java.version>25</java.version>
<vaadin.version>25.2.3</vaadin.version>
```

`Tested in Chome, and Mozilla, on Linux Ubuntu 24.04.3` 

But something strange is happening—do I need to click twice for it to work, or am I just imagining things? 

- https://vaadin.com/forum/t/clipboard-copy/164697/14
- https://github.com/rucko24/xsd-validator-ui/blob/feature/clipboard-native-api/src/main/java/com/rubn/xsdvalidator/view/Input.java#L279

---

In the main branch, there is the direct clipboard with a JsModule that is working

- https://github.com/rucko24/xsd-validator-ui/blob/main/src/main/resources/META-INF/resources/frontend/scripts/copy_to_clipboard.js

**Comment by Artur-** [comment]
You are setting up the clipboard click handler (which runs in the browser) in the server side click event. So the first click event reaches the server, it sets up a copy to clipboard handler for the item and nothing more. The second click event copies to clipboard and also goes to the server and sets up another copy to clipboard handler for the item. The third click then copies two times to the clipboard etc
